src/main/java/at/hannibal2/skyhanni/features/misc/GhostCounter.kt

```diff
@@ -0,0 +1,326 @@
+        )
+
+        for ((text, value) in list) {
+            val (total, session) = value
+            val se = if(session == "") "" else "($session)"
+            addAsSingletonList(config.formatText.replace("&", "§")
+                    .replace("%text%", text)
+                    .replace("%value%", "$total")
+                    .replace("%session%", se))
+        }
+    }
+
+    @SubscribeEvent
+    fun onInventoryOpen(event: InventoryOpenEvent) {
+        if (!isEnabled()) return
+        val inventoryName = event.inventoryName
+        if (inventoryName.startsWith("Accessory Bag")) {
+            val stacks = event.inventoryItems
+            for ((_, stack) in stacks) {
+                val name = stack.name ?: return
+                if (name.removeColor().contains("Scavenger Talisman")) {
+                    hasScavengerTalisman = true
+                }
+            }
+        }
+    }
+
+    // Part of this was taken from GhostCounterV3 CT module
+    @SubscribeEvent
+    fun onActionBar(event: LorenzActionBarEvent) {
+        if (!isEnabled()) return
+        skillXPPattern.matchMatcher(event.message) {
+            val gained = group("gained").toDouble()
+            val total = group("total")
+            if (total != lastXp) {
+                val res = total.replace("\\D".toRegex(), "")
+                gain = (res.toLong() - lastXp.toLong()).toDouble().roundToInt()
+                num = (gain.toDouble() / gained)
+                if (gained in 150.0..450.0) {
+                    if (lastXp != "0") {
+                        if (num >= 0) {
+                            KILLS.add(num)
+                            GHOSTSINCESORROW.add(num)
+                            InventoryUtils.getItemInHand()?.let {
+                                val scavengerLevel = it.getEnchantments()?.get("scavenger") ?: 0
+                                var baseValue = 0.0
+                                baseValue += (scavengerMap.getOrDefault(scavengerLevel, 0).toDouble() * 250)
+                                if (hasScavengerTalisman || config.forceScavengerTalisman)
+                                    baseValue += (250 * 0.5)
+                                baseValue += KILLCOMBOCOINS.getInt() * num
+                                val scav = (baseValue * num.roundToInt()).roundToInt().toDouble()
+                                SCAVENGERCOINS.add(scav)
+                                KILLCOMBO.add(num)
+
+                                session.kills += num
+                                session.scavenger+= scav
+                            }
+                        }
+                    }
+                }
+                lastXp = res
+            }
+        }
+    }
+
+    @SubscribeEvent
+    fun onChat(event: LorenzChatEvent) {
+        if (!isEnabled()) return
+        if (LorenzUtils.skyBlockIsland != IslandType.DWARVEN_MINES) return
+        sorrowPattern.matchMatcher(event.message) {
```

Comment:
Maybe move the pattern into the enum and iterate over all enum entires here instead

Reply:
Like
enum class Option(val save: Boolean = true, pattern: Pattern = "".toPattern()) {
KILLS("pattern here"),
...
}

some enum would have no pattern, what would happen when i do

for (opt in Option.values()) {
    opt.pattern.matchMatcher(event.message){
        opt.set(1.0)
    }
}

sometime i need to set other Option value so it's either me not understanding what u say or can't figure out how to do it

Reply:
Remove these 3 elements from the Option enum:
`BESTIARY_NEXTLEVEL`
`BESTIARY_CURRENTKILL`
`BESTIARY_KILLNEEDED`
Save them separately in GhostCounterStorage

Remove all SESSION elements from Option enum.
Use instead 2 instances of Option, one for the current session, one for total data.

Since GhostCounterStorage increases the Misc.java class size from

Reply:
Maybe handle `TOTALMF` also separately?

Reply:
You can loop over the enum `Option` in line 173 as well